network/tcp.go

```diff
@@ -0,0 +1,422 @@
+		// read the size of the next packet
+		n, err := c.conn.Read(b)
+		// if error then quit
+		if err != nil {
+
+			return nil, handleError(err)
+		}
+		// put it in the longterm buffer
+		if _, err := buffer.Write(b[:n]); err != nil {
+			log.Error("Couldn't write to buffer:", err)
+		}
+		read += Size(n)
+		b = b[n:]
+	}
+
+	// set the size read
+	c.updateRx(uint64(read))
+	return buffer.Bytes(), nil
+}
+
+// how many bytes do we write at once on the socket
+// 1400 seems a safe choice regarding the size of a ethernet packet.
+// https://stackoverflow.com/questions/2613734/maximum-packet-size-for-a-tcp-connection
+const maxChunkSize Size = 1400
+
+// Send will convert the NetworkMessage into an ApplicationMessage
+// and send it with send()
+// Returns an error if anything was wrong
+func (c *TCPConn) Send(ctx context.Context, obj Body) error {
+	c.sendMutex.Lock()
+	defer c.sendMutex.Unlock()
+	am, err := NewNetworkPacket(obj)
+	if err != nil {
+		return fmt.Errorf("Error converting packet: %v\n", err)
+	}
+	log.Lvlf5("Message SEND => %+v", am)
+	var b []byte
+	b, err = am.MarshalBinary()
+	if err != nil {
+		return fmt.Errorf("Error marshaling  message: %s", err.Error())
+	}
+	return c.send(b)
+}
+
+// send takes care of sending this slice of bytes FULLY to the connection
+func (c *TCPConn) send(b []byte) error {
+	// First write the size
+	packetSize := Size(len(b))
+	if err := binary.Write(c.conn, globalOrder, packetSize); err != nil {
+		return err
+	}
+	// Then send everything through the connection
+	// Send chunk by chunk
+	var sent Size
+	for sent < packetSize {
+		length := packetSize - sent
+		if length > maxChunkSize {
+			length = maxChunkSize
+		}
+
+		// Sending 'length' bytes
+		log.Lvl4("Sending from", c.conn.LocalAddr(), "to", c.conn.RemoteAddr())
+		n, err := c.conn.Write(b[:length])
+		if err != nil {
+			log.Error("Couldn't write chunk starting at", sent, "size", length, err)
+			return handleError(err)
+		}
+		sent += Size(n)
+		log.Lvl5("Sent", sent, "out of", packetSize)
+
+		// bytes left to send
+		b = b[n:]
+	}
+	// update stats on the connection
+	c.updateTx(uint64(packetSize))
+	return nil
+}
+
+// Remote returns the name of the peer at the end point of
+// the connection
+func (c *TCPConn) Remote() Address {
+	return c.endpoint
+}
+
+// Local returns the local address and port
+func (c *TCPConn) Local() Address {
+	return NewTCPAddress(c.conn.LocalAddr().String())
+}
+
+// Type implements the Conn interface
+func (c *TCPConn) Type() ConnType {
+	return PlainTCP
+}
+
+// Close ... closes the connection
+func (c *TCPConn) Close() error {
+	c.closedMut.Lock()
+	defer c.closedMut.Unlock()
+	if c.closed == true {
+		return nil
+	}
+	err := c.conn.Close()
+	c.closed = true
+	if err != nil {
+		return handleError(err)
+	}
+	return nil
+}
+
+// handleError produces the higher layer error depending on the type
+// so user of the package can know what is the cause of the problem
+func handleError(err error) error {
```

Comment:
Look where I use `ErrClosed`:
```
network/local.go|129 col 10| return ErrClosed
network/local.go|238 col 10| return ErrClosed
network/local.go|287 col 16| return nil, ErrClosed
network/local.go|292 col 15| return nil, ErrClosed
network/router.go|181 col 14| if err == ErrClosed || err == ErrEOF || err == ErrTemp {
network/tcp.go|192 col 10| return ErrClosed
network/tcp.go|207 col 10| return ErrClosed
network/tcp.go|319 col 27| if handleError(err) != ErrClosed {
network/local_test.go|100 col 18| assert.Equal(t, ErrClosed, err)
network/local_test.go|101 col 18| assert.Equal(t, ErrClosed, outgoing.Close())
network/local_test.go|276 col 12| if err != ErrClosed {
network/local_test.go|279 col 18| assert.Equal(t, ErrClosed, outgoing.Close())
```
So it's used quite a lot already. 
I would not ask the question regarding the difficulty, but rather what are the pros/cons of each side ? 
Pros:
* Unified error handling: Having a LocalConn which can return errors that have **the same meaning** (and **are** indeed the same) as with TCPConn, so router use them independently, I find that rather useful.
* Simpler error handling at higher level: 
`if strings.Contains(err.Error(), "Closed") ` vs `if err == ErrClosed` ... 
The first one is error prone and not clear.
Cons:
* We can't have the exact error string. However I coded theses errors while looking at the source code of golang net so I tried to catch most of them. If one is not known, then it can returns this error.